src/vault.rs

```diff
@@ -1,61 +1,349 @@
+        // `env::block_height()`, and in this case, it would be possible for us
+        // to miss some rewards if they were distributed in the same block but
+        // after this function call.
+        if start_from_block_height >= until_block_height {
+            return;
+        }
+
+        let mut accumulated_fees_in_span = 0;
+        let mut last_block_height = start_from_block_height;
+
+        for (block_height, fees) in self
+            .collateral_asset_fee_distribution_log
+            .iter_from(start_from_block_height)
+        {
+            if block_height >= until_block_height {
+                break;
+            }
+
+            let total_loan_asset_deposited_at_distribution = self
+                .total_loan_asset_deposited_log
+                .get(
+                    &self
+                        .total_loan_asset_deposited_log
+                        .floor_key(&block_height)
+                        .unwrap(),
+                )
+                .unwrap();
+
+            // this discards fractional fees
+            let portion_of_fees = fees
+                .checked_mul(lender_entry.loan_asset_deposited)
+                .unwrap()
+                .checked_div(total_loan_asset_deposited_at_distribution)
+                .unwrap();
+
+            accumulated_fees_in_span += portion_of_fees;
+
+            last_block_height = block_height;
+        }
+
+        lender_entry.collateral_asset_fees_earned += accumulated_fees_in_span;
+        lender_entry.collateral_asset_fee_distribution_updated_until_block_height =
+            last_block_height;
+    }
+
+    pub fn can_borrower_be_liquidated(
+        &self,
+        borrower_id: &AccountId,
+        collateral_asset_multiplier: (u128, u128),
+        loan_asset_multiplier: (u128, u128),
+    ) -> bool {
+        let Some(borrower_entry) = self.borrowers.get(borrower_id) else {
+            return false;
+        };
+
+        let scaled_collateral_value = borrower_entry.collateral_asset_deposited
+            * collateral_asset_multiplier.0
+            * loan_asset_multiplier.1
+            * self.min_collateral_ratio.1 as u128;
+        let scaled_loan_value = borrower_entry.loan_asset_borrowed
+            * loan_asset_multiplier.0
+            * collateral_asset_multiplier.1
+            * self.min_collateral_ratio.0 as u128;
+
+        scaled_loan_value > scaled_collateral_value
+    }
+
+    pub fn record_liquidation(&mut self, borrower_id: &AccountId) {
+        let mut borrower_entry = self.borrowers.get(borrower_id).unwrap_or_default();
+
+        let liquidated_collateral = borrower_entry.collateral_asset_deposited;
+        borrower_entry.collateral_asset_deposited = 0;
+        let liquidated_loan = borrower_entry.loan_asset_borrowed;
+        borrower_entry.loan_asset_borrowed = 0;
+        // TODO: Do we distribute the liquidated collateral as fees/rewards?
+        // TODO: Do we swap the liqidated funds to the loan asset?
```

Comment:
The liquidator will be selling the collateral asset into the loan asset. They will take care of that for us and then collect the liquidator reward. We need to think through how the liquidation process occurs in detail to prevent the liquidators from keeping all the collateral assets. There should be some sort of liquidation function that handles this

Reply:
I'm open to the contract being the liquidator to start, but if you have another idea that you think would work better, I'm open to that too.